### PR TITLE
sched/group: change type of task group member to single queue

### PR DESCRIPTION
**Pull request by anchao**
## Summary

sched/group: change type of task group member to single queue

Change the type of task group member to single list chain to avoid accessing the memory allocator to improve the performance

Signed-off-by: chao an <anchao@lixiang.com>

## Impact

N/A

## Testing

ci-check

### CODE REVIEW

File: sched/task/task_exithook.c [hunk 1]
```diff
@@ -174,7 +174,7 @@ static inline void nxtask_sigchild(pid_t ppid, FAR struct tcb_s *ctcb,
    * should generate SIGCHLD.
    */
 
-  if (chgrp->tg_nmembers == 1)
+  if (sq_count(&chgrp->tg_members) == 1)
```
Comment (xiaoxiang781216) [inline sched/task/task_exithook.c:177]: too slow, let's add sq_is_singular?

File: sched/group/group_leave.c [hunk 2]
```diff
@@ -226,17 +166,17 @@ void group_leave(FAR struct tcb_s *tcb)
...
-      group_removemember(group, tcb->pid);
+#ifdef HAVE_GROUP_MEMBERS
+      flags = spin_lock_irqsave_wo_note(NULL);
+      sq_rem(&tcb->member, &group->tg_members);
+      spin_unlock_irqrestore_wo_note(NULL, flags);
 
       /* Have all of the members left the group? */
 
-      if (group->tg_nmembers == 0)
+      if (sq_count(&group->tg_members) == 0)
```
Comment (xiaoxiang781216) [inline sched/group/group_leave.c:178]: call sq_is_empty

File: sched/group/group_killchildren.c [hunk 3]
```diff
@@ -191,7 +191,7 @@ int group_kill_children(FAR struct tcb_s *tcb)
   ret = CONFIG_GROUP_KILL_CHILDREN_TIMEOUT_MS;
   while (1)
     {
-      if (tcb->group->tg_nmembers <= 1)
+      if (sq_count(&tcb->group->tg_members) <= 1)
```
Comment (xiaoxiang781216) [inline sched/group/group_killchildren.c:194]: sq_is_empty

File: sched/task/task_exithook.c (hunk 1 above)
Comment (anchao) [reply sched/task/task_exithook.c]: Done

File: include/nuttx/list.h [hunk 4]
```diff
@@ -348,4 +348,10 @@ static inline size_t list_length(FAR struct list_node *list)
   return cnt;
 }
 
+static inline int
+list_is_singular(FAR const struct list_node *list)
```
Comment (xiaoxiang781216) [inline include/nuttx/list.h:352]: and move after line 74:
```
#define list_is_singular(list) ((list)->next == (list)->prev)
```

File: sched/group/group.h [hunk 5]
```diff
@@ -58,9 +58,12 @@ void task_initialize(void);
 
 int  group_allocate(FAR struct task_tcb_s *tcb, uint8_t ttype);
 void group_initialize(FAR struct task_tcb_s *tcb);
-#ifndef CONFIG_DISABLE_PTHREAD
+#ifdef HAVE_GROUP_MEMBERS
 int  group_bind(FAR struct pthread_tcb_s *tcb);
 int  group_join(FAR struct pthread_tcb_s *tcb);
+#else
+# define group_bind(tcb) OK
```
Comment (xiaoxiang781216) [inline sched/group/group.h:65]:
```suggestion
#  define group_bind(tcb) OK
```

File: include/nuttx/list.h (hunk 4 above)
Comment (anchao) [reply include/nuttx/list.h]: Done

File: sched/group/group.h [hunk 6]
```diff
@@ -58,9 +58,12 @@ void task_initialize(void);
 
 int  group_allocate(FAR struct task_tcb_s *tcb, uint8_t ttype);
 void group_initialize(FAR struct task_tcb_s *tcb);
-#ifndef CONFIG_DISABLE_PTHREAD
+#ifdef HAVE_GROUP_MEMBERS
 int  group_bind(FAR struct pthread_tcb_s *tcb);
 int  group_join(FAR struct pthread_tcb_s *tcb);
+#else
+#  define group_bind(tcb) OK
```
Comment (xiaoxiang781216) [inline sched/group/group.h:65]: why need add empty macro?

File: sched/group/group.h [hunk 7]
```diff
@@ -58,9 +58,12 @@ void task_initialize(void);
 
 int  group_allocate(FAR struct task_tcb_s *tcb, uint8_t ttype);
 void group_initialize(FAR struct task_tcb_s *tcb);
-#ifndef CONFIG_DISABLE_PTHREAD
+#ifdef HAVE_GROUP_MEMBERS
```
Comment (xiaoxiang781216) [inline sched/group/group.h:61]: why need change

File: sched/group/group.h (hunk 6 above)
Comment (anchao) [reply sched/group/group.h:65]: Done

File: sched/group/group.h (hunk 7 above)
Comment (anchao) [reply sched/group/group.h]: Done

File: fs/procfs/fs_procfsproc.c [hunk 8]
```diff
@@ -1202,10 +1205,12 @@ static ssize_t proc_groupstatus(FAR struct proc_file_s *procfile,
       return totalsize;
     }
 
-  for (i = 0; i < group->tg_nmembers; i++)
+  sq_for_every_safe(&group->tg_members, curr, next)
     {
+      FAR struct tcb_s *mtcb =
```
Comment (xiaoxiang781216) [inline fs/procfs/fs_procfsproc.c:1210]: can reuse tcb?

File: sched/clock/clock_gettime.c [hunk 9]
```diff
@@ -183,20 +188,24 @@ int clock_gettime(clockid_t clock_id, struct timespec *tp)
...
-          for (i = group->tg_nmembers - 1; i >= 0; i--)
+          flags = spin_lock_irqsave_wo_note(NULL);
+          sq_for_every_safe(&group->tg_members, curr, next)
             {
-              tcb = nxsched_get_tcb(group->tg_members[i]);
-              if (tcb != NULL)
-                {
-                  runtime += tcb->run_time;
-                }
+              FAR struct tcb_s *mtcb =
```
Comment (xiaoxiang781216) [inline sched/clock/clock_gettime.c:198]: could reuse tcb too?

File: sched/clock/clock_gettime.c [hunk 10]
```diff
@@ -183,20 +188,24 @@ int clock_gettime(clockid_t clock_id, struct timespec *tp)
 
       if (tcb != NULL)
         {
+# ifdef HAVE_GROUP_MEMBERS
           group = tcb->group;
           runtime = 0;
 
-          flags = enter_critical_section();
-          for (i = group->tg_nmembers - 1; i >= 0; i--)
+          flags = spin_lock_irqsave_wo_note(NULL);
```
Comment (xiaoxiang781216) [inline sched/clock/clock_gettime.c:195]: spin_lock_irqsave

File: sched/group/group_foreachchild.c [hunk 11]
```diff
@@ -58,23 +60,27 @@
...
   DEBUGASSERT(group);
 
   /* Visit the main thread last (if present) */
 
-  for (i = group->tg_nmembers - 1; i >= 0; i--)
+  sq_for_every_safe(&group->tg_members, curr, next)
     {
-      ret = handler(group->tg_members[i], arg);
-      if (ret != 0)
+      FAR struct tcb_s *mtcb =
```
Comment (xiaoxiang781216) [inline sched/group/group_foreachchild.c:73]: mtcb->tcb

File: sched/group/group_join.c [hunk 12]
```diff
@@ -203,32 +97,19 @@ int group_bind(FAR struct pthread_tcb_s *tcb)
...
-  if (ret < 0)
-    {
-      return ret;
-    }
-#else
-  flags = spin_lock_irqsave(NULL);
-  group->tg_nmembers++;
-  spin_unlock_irqrestore(NULL, flags);
-#endif
+  flags = spin_lock_irqsave_wo_note(NULL);
```
Comment (xiaoxiang781216) [inline sched/group/group_join.c:110]: spin_lock_irqsave

File: sched/group/group_leave.c [hunk 13]
```diff
@@ -226,17 +166,17 @@ void group_leave(FAR struct tcb_s *tcb)
...
-      /* Remove the member from group.  This function may be called
-       * during certain error handling before the PID has been
-       * added to the group.  In this case tcb->pid will be uninitialized
-       * group_removemember() will fail.
-       */
+      /* Remove the member from group. */
 
-      group_removemember(group, tcb->pid);
+#ifdef HAVE_GROUP_MEMBERS
+      flags = spin_lock_irqsave_wo_note(NULL);
```
Comment (xiaoxiang781216) [inline sched/group/group_leave.c:172]: spin_lock_irqsave

File: fs/procfs/fs_procfsproc.c (hunk 8 above)
Comment (anchao) [reply fs/procfs/fs_procfsproc.c]: Done

File: sched/clock/clock_gettime.c (hunk 9 above)
Comment (anchao) [reply sched/clock/clock_gettime.c]: Done

File: sched/clock/clock_gettime.c (hunk 10 above)
Comment (anchao) [reply sched/clock/clock_gettime.c]: Done

File: sched/group/group_foreachchild.c (hunk 11 above)
Comment (anchao) [reply sched/group/group_foreachchild.c:73]: Done

File: sched/group/group_join.c (hunk 12 above)
Comment (anchao) [reply sched/group/group_join.c]: Done

File: sched/group/group_leave.c (hunk 13 above)
Comment (anchao) [reply sched/group/group_leave.c]: Done

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by masayuki2009** [comment]
@anchao 
I noticed that spresense:elf and sabre-6quad:elf (qemu) stops with this PR.

```
/home/ishikawa/opensource/QEMU/qemu-8.2.0/build/qemu-system-arm -semihosting -M sabrelite -m 1024 -smp 4 -kernel nuttx -nographic
Initial memory usage: 6228
Registering romdisk at /dev/ram0

Memory Usage after romdisk_register:
  Before:     6228 After:     6284 Change:       56
Mounting ROMFS filesystem at target=/mnt/romfs with source=/dev/ram0

Memory Usage after mount:
  Before:     6284 After:     6844 Change:      560

****************************************************************************
* Executing errno
****************************************************************************


Memory Usage after exec:
  Before:     6844 After:    10236 Change:     3392
Wait a bit for test completion
Hello, World on stdout
Hello, World on stderr
We failed to open "aflav-sautga-ay!" errno is 2
```

**Comment by masayuki2009** [comment]
@anchao 
If I revert the second commit, it works.

```
/home/ishikawa/opensource/QEMU/qemu-8.2.0/build/qemu-system-arm -semihosting -M sabrelite -m 1024 -smp 4 -kernel nuttx -nographic
Initial memory usage: 6252
Registering romdisk at /dev/ram0

Memory Usage after romdisk_register:
  Before:     6252 After:     6308 Change:       56
Mounting ROMFS filesystem at target=/mnt/romfs with source=/dev/ram0

Memory Usage after mount:
  Before:     6308 After:     6868 Change:      560

****************************************************************************
* Executing errno
****************************************************************************


Memory Usage after exec:
  Before:     6868 After:    10268 Change:     3400
Wait a bit for test completion
Hello, World on stdout
Hello, World on stderr
We failed to open "aflav-sautga-ay!" errno is 2

Memory Usage after program execution:
  Before:    10268 After:     9100 Change:    -1168

****************************************************************************
* Executing hello
****************************************************************************


Memory Usage after exec:
  Before:     9100 After:    10284 Change:     1184
Wait a bit for test completion
Getting ready to say "Hello, world"

Hello, world!
It has been said.

argc	= 1
argv	= 0x1082a838
argv[0]	= (0x1082a840) "hello"
argv[1]	= 0
Goodbye, world!

Memory Usage after program execution:
  Before:    10284 After:     9100 Change:    -1184

****************************************************************************
* Executing hello++1
****************************************************************************


Memory Usage after exec:
  Before:     9100 After:    10028 Change:      928
Wait a bit for test completion
Hello, World!

Memory Usage after program execution:
  Before:    10028 After:     9100 Change:     -928

****************************************************************************
* Executing hello++2
****************************************************************************


Memory Usage after exec:
  Before:     9100 After:    10596 Change:     1496
Wait a bit for test completion
main: Started.  Creating MyThingSayer
CThingSayer::CThingSayer: I am!
main: Created MyThingSayer=0x1082a4c0
main: Calling MyThingSayer->Initialize
CThingSayer::Initialize: When told, I will say 'Hello, World!'
main: Calling MyThingSayer->SayThing
CThingSayer::SayThing: I am now saying 'Hello, World!'
main: Destroying MyThingSayer
CThingSayer::~CThingSayer: I cease to be
CThingSayer::~CThingSayer: I will never say 'Hello, World!' again
main: Returning

Memory Usage after program execution:
  Before:    10596 After:     9100 Change:    -1496

****************************************************************************
* Executing hello++3
****************************************************************************


Memory Usage after exec:
  Before:     9100 After:    10708 Change:     1608
Wait a bit for test completion
CThingSayer::CThingSayer: I am!
main: Started.  MyThingSayer should already exist
main: Calling MyThingSayer.Initialize
CThingSayer::Initialize: When told, I will say 'Hello, World!'
main: Calling MyThingSayer.SayThing
CThingSayer::SayThing: I am now saying 'Hello, World!'
main: Returning.  MyThingSayer should be destroyed

Memory Usage after program execution:
  Before:    10708 After:     9100 Change:    -1608

****************************************************************************
* Executing mutex
****************************************************************************


Memory Usage after exec:
  Before:     9100 After:    10740 Change:     1640
Wait a bit for test completion
Starting thread 1
Starting thread 2
Stopping threads
	Thread1	Thread2
Loops	10	10
Errors	0	0

Memory Usage after program execution:
  Before:    10740 After:    11356 Change:      616

****************************************************************************
* Executing pthread
****************************************************************************


Memory Usage after exec:
  Before:    11356 After:    10628 Change:     -728
Wait a bit for test completion
PARENT: started

PARENT: calling pthread_start with arg=305419896
CHILD: started with arg=305419896
CHILD: returning -2023406815
PARENT child exitted with -2023406815
PARENT returning success


Memory Usage after program execution:
  Before:    10628 After:    11356 Change:      728

****************************************************************************
* Executing signal
****************************************************************************


Memory Usage after exec:
  Before:    11356 After:    11732 Change:      376
Wait a bit for test completion
Setting up signal handlers from pid=12
Old SIGUSR1 sighandler at 0
New SIGUSR1 sighandler at 0x1082a921
Old SIGUSR2 sighandler at 0
New SIGUSR2 sighandler at 0x1082a921
Raising SIGUSR1 from pid=12
Kill-ing SIGUSR1 from pid=12
siguser_action: Received signo=10 siginfo=0x10828b28 arg=0
  SIGUSR1 received
siginfo:
  si_signo  = 10
  si_code   = 0
  si_errno  = 4
  si_value  = 0
SIGUSR1 raised from pid=12
SIGUSR1 not received

Memory Usage after program execution:
  Before:    11732 After:     9236 Change:    -2496

****************************************************************************
* Executing struct
****************************************************************************


Memory Usage after exec:
  Before:     9236 After:    10668 Change:     1432
Wait a bit for test completion
Calling getstruct()
getstruct returned 0x1082a74c
  n = 42 (vs 42) PASS
  pn = 0x10829fb8 (vs 0x10829fb8) PASS
 *pn = 87 (vs 87) PASS
  ps = 0x1082a748 (vs 0x1082a748) PASS
  ps->n = 117 (vs 117) PASS
  pf = 0x1082a549 (vs 0x1082a549) PASS
Calling mystruct->pf()
In dummyfunc() -- PASS
Exit-ing

Memory Usage after program execution:
  Before:    10668 After:     9236 Change:    -1432

****************************************************************************
* Executing task
****************************************************************************


Memory Usage after exec:
  Before:     9236 After:    10980 Change:     1744
Wait a bit for test completion
Parent: Started, pid=15
Parent: Calling task_create()
Parent: Waiting for child (pid=18)
Child: execv was successful!
Child: argc=2
Child: argv[0]="child"
Child: argv[1]="Hello from your parent!"
Child: Exit-ting with status=0
Parent: Exit-ing

Memory Usage after program execution:
  Before:    10980 After:    11468 Change:      488

Memory Usage End-of-Test:
```

**Comment by masayuki2009** [comment]
@anchao 
sim:elf (on ubuntu 20.04 x86_64) crashed.

```
NuttShell (NSH) NuttX-12.0.0
MOTD: username=admin password=Administrator
nsh> elf
Initial memory usage: 1332584
elf_main: Registering romdisk at /dev/ram6

Memory Usage after romdisk_register:
  Before:  1332584 After:  1332712 Change:      128
elf_main: Mounting ROMFS filesystem at target=/mnt/romfs with source=/dev/ram6

Memory Usage after mount:
  Before:  1332712 After:  1333448 Change:      736
testheader:
****************************************************************************
* Executing errno
****************************************************************************


Memory Usage after exec:
  Before:  1333448 After:  1402776 Change:    69328
elf_main: Wait a bit for test completion
Hello, World on stderr
We failed to open "aflav-sautga-ay!" errno is 2
Hello, World on stdout
_assert: Current Version: NuttX  12.0.0 ec08031e4b Mar  7 2024 15:30:28 sim
_assert: Assertion failed : at file: sched/sched_addreadytorun.c:98 task: nsh_main process: nsh_main 0x42c376
Segmentation fault (core dumped)
```

**Comment by masayuki2009** [comment]
@anchao 
May I revert this PR until you fix the above errors?

**Comment by anchao** [comment]
> @anchao May I revert this PR until you fix the above errors?

@masayuki2009 sorry for regression, I will take a look at this issue as soon as possible.

**Pull request merged**